File: src/session/SessionManager.ts
import { logger } from '../utils/logging.js';

export interface SessionContext {
  id: string;
  messages: Array<{
    role: 'system' | 'user' | 'assistant';
    content: string;
    timestamp: number;
  }>;
  createdAt: number;
  lastAccessedAt: number;
}

export class SessionManager {
  private sessions: Map<string, SessionContext> = new Map();
  private readonly maxSessions = 100;
  private readonly sessionTimeoutMs = 30 * 60 * 1000; // 30 minutes

  private readonly systemPrompt = `You are JARVIS, a personal AI assistant. You are helpful, concise, and professional.
Never reference Tony Stark, Iron Man, Marvel characters, or fictional scenarios. You assist real users with real tasks.
Keep responses brief and focused. Remember user preferences and names when told.`;

  getOrCreate(sessionId: string): SessionContext {
    let session = this.sessions.get(sessionId);
    
    if (!session) {
      session = {
        id: sessionId,
        messages: [
          {
            role: 'system',
            content: this.systemPrompt,
            timestamp: Date.now()
          }
        ],
        createdAt: Date.now(),
        lastAccessedAt: Date.now()
      };
      
      this.sessions.set(sessionId, session);
      logger.debug(`Created new session: ${sessionId}`);
      
      // Cleanup old sessions if we're at the limit
      this.cleanup();
    } else {
      session.lastAccessedAt = Date.now();
    }
    
    return session;
  }

  appendUser(sessionId: string, content: string): SessionContext {
    const session = this.getOrCreate(sessionId);
    
    session.messages.push({
      role: 'user',
      content,
      timestamp: Date.now()
    });
    
    return session;
  }

  appendAssistant(sessionId: string, content: string): SessionContext {
    const session = this.getOrCreate(sessionId);
    
    session.messages.push({
      role: 'assistant',
      content,
      timestamp: Date.now()
    });
    
    return session;
  }

  getMessages(sessionId: string): SessionContext['messages'] {
    const session = this.getOrCreate(sessionId);
    return session.messages;
  }

  reset(sessionId: string): void {
    this.sessions.delete(sessionId);
    logger.debug(`Reset session: ${sessionId}`);
  }

  private cleanup(): void {
    if (this.sessions.size <= this.maxSessions) return;
    
    const now = Date.now();
    const sessionsToDelete: string[] = [];
    
    // Find expired sessions
    for (const [id, session] of this.sessions) {
      if (now - session.lastAccessedAt > this.sessionTimeoutMs) {
        sessionsToDelete.push(id);
      }
    }
    
    // If we still have too many, delete the oldest
    if (this.sessions.size - sessionsToDelete.length > this.maxSessions) {
      const sortedSessions = Array.from(this.sessions.entries())
        .sort((a, b) => a[1].lastAccessedAt - b[1].lastAccessedAt);
      
      const excess = this.sessions.size - sessionsToDelete.length - this.maxSessions;
      for (let i = 0; i < excess; i++) {
        sessionsToDelete.push(sortedSessions[i][0]);
      }
    }
    
    // Delete sessions
    sessionsToDelete.forEach(id => {
      this.sessions.delete(id);
      logger.debug(`Cleaned up session: ${id}`);
    });
  }

  getSessions(): string[] {
    return Array.from(this.sessions.keys());
  }

  getSessionCount(): number {
    return this.sessions.size;
  }
}

// Global session storage for llama.cpp contexts and chats
const llamaSessions = new Map<string, { context: any; chat: any }>();

// Get or create persistent context & chat for a session
export async function getSession<TCtx, TChat>(
  sessionId: string,
  createCtx: () => Promise<TCtx> | TCtx,
  createChat: (ctx: TCtx) => Promise<TChat> | TChat
): Promise<{ context: TCtx; chat: TChat }> {
  if (llamaSessions.has(sessionId)) {
    const existing = llamaSessions.get(sessionId)!;
    return { context: existing.context, chat: existing.chat };
  }

  logger.debug(`Creating new llama session: ${sessionId}`);
  const context = await createCtx();
  const chat = await createChat(context);
  
  llamaSessions.set(sessionId, { context, chat });
  
  // Cleanup old sessions if we have too many
  if (llamaSessions.size > 50) {
    const oldestKey = llamaSessions.keys().next().value;
    if (oldestKey) {
      llamaSessions.delete(oldestKey);
      logger.debug(`Cleaned up old llama session: ${oldestKey}`);
    }
  }
  
  return { context, chat };
}

File: jarvis/session/session_manager.py
"""Ported from src/session/SessionManager.ts."""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Literal, Optional

from ..utils.logging import logger

Role = Literal["system", "user", "assistant"]

SYSTEM_PROMPT = (
    "You are JARVIS, a personal AI assistant. You are helpful, concise, and professional.\n"
    "Never reference Tony Stark, Iron Man, Marvel characters, or fictional scenarios. "
    "You assist real users with real tasks.\n"
    "Keep responses brief and focused. Remember user preferences and names when told."
)


@dataclass
class Message:
    role: Role
    content: str
    timestamp: float


@dataclass
class SessionContext:
    id: str
    messages: list[Message] = field(default_factory=list)
    created_at: float = field(default_factory=lambda: time.time() * 1000)
    last_accessed_at: float = field(default_factory=lambda: time.time() * 1000)


class SessionManager:
    MAX_SESSIONS = 100
    SESSION_TIMEOUT_MS = 30 * 60 * 1000

    def __init__(self) -> None:
        self._sessions: dict[str, SessionContext] = {}

    def get_or_create(self, session_id: str) -> SessionContext:
        session = self._sessions.get(session_id)
        if session is None:
            session = SessionContext(id=session_id, messages=[Message("system", SYSTEM_PROMPT, time.time() * 1000)])
            self._sessions[session_id] = session
            logger.debug(f"Created new session: {session_id}")
            self._cleanup()
        else:
            session.last_accessed_at = time.time() * 1000
        return session

    def append_user(self, session_id: str, content: str) -> SessionContext:
        session = self.get_or_create(session_id)
        session.messages.append(Message("user", content, time.time() * 1000))
        return session

    def append_assistant(self, session_id: str, content: str) -> SessionContext:
        session = self.get_or_create(session_id)
        session.messages.append(Message("assistant", content, time.time() * 1000))
        return session

    def get_messages(self, session_id: str) -> list[Message]:
        return self.get_or_create(session_id).messages

    def reset(self, session_id: str) -> None:
        self._sessions.pop(session_id, None)
        logger.debug(f"Reset session: {session_id}")

    def _cleanup(self) -> None:
        if len(self._sessions) <= self.MAX_SESSIONS:
            return

        now = time.time() * 1000
        to_delete = [sid for sid, s in self._sessions.items() if now - s.last_accessed_at > self.SESSION_TIMEOUT_MS]

        remaining = len(self._sessions) - len(to_delete)
        if remaining > self.MAX_SESSIONS:
            oldest = sorted(self._sessions.items(), key=lambda kv: kv[1].last_accessed_at)
            excess = remaining - self.MAX_SESSIONS
            to_delete.extend(sid for sid, _ in oldest[:excess])

        for sid in to_delete:
            self._sessions.pop(sid, None)
            logger.debug(f"Cleaned up session: {sid}")

    def get_sessions(self) -> list[str]:
        return list(self._sessions.keys())

    def get_session_count(self) -> int:
        return len(self._sessions)


# Global context/chat storage for llama.cpp sessions (KV-cache reuse),
# mirroring the module-level `llamaSessions` map in SessionManager.ts.
_llama_sessions: dict[str, Any] = {}


def get_llama_session(session_id: str, create: Any) -> Any:
    if session_id in _llama_sessions:
        return _llama_sessions[session_id]

    logger.debug(f"Creating new llama session: {session_id}")
    session = create()
    _llama_sessions[session_id] = session

    if len(_llama_sessions) > 50:
        oldest_key = next(iter(_llama_sessions))
        del _llama_sessions[oldest_key]
        logger.debug(f"Cleaned up old llama session: {oldest_key}")

    return session
